Add storage module from google-cloud-cpp to build system

pamarcos opened the pull request:
Preparation work for adding native GCS table function/engine

Add storage module from google-cloud-cpp to build system:

- Clean CMake files to avoid so much commented code. WIN32 and unnecessary installation steps removed
- Add nlohmann's JSON dependency to contrib
- Add crc32c dependency to contrib. Seems everyone and their cousing integrates their own version of crc32 somehow. We have it within ClickHouse in rocksdb, arrow, libhdfs3, librdkafka and abseil. However, I didn't think it was clean to use a dependency of a dependency to build yet another dependency. I did try to use the abseil's one, though, but it required changes in google-cloud-cpp, so rather than modifying our fork, I thought it'd be simpler to maintain to just use google's take on crc32c.
- Split `ENABLE_GOOGLE_CLOUD_CPP` for Storage (and others that may come in the future) and `ENABLE_GOOGLE_CLOUD_CPP_KMS` for `KMS` which is only used in private
- Add `ci-build` label support to avoid running all tests. Reasoning is that in order to build darwin and freebsd, all tests are blocking them and sometimes (as in this case) we are only interested in builds

### Check that gcloud_storage example builds and links for Linux, macOS and FreeBSD for archs x86_64 and aarch64

I added to `build_clickhouse.py` the `DENABLE_EXAMPLES` for all flavors to ensure they work ok. In macOS and FreeBSD the field example builds linking, but it's unrelated to these changes.

https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=82881&sha=73242ae58fc8157e4b60379d2f4bce2576537360&name_0=PR

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add storage module from google-cloud-cpp to build system

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

pamarcos committed:
Add Storage from google-cloud-cpp along with its dependencies

- nlohmann's json
- crc32c

Also silence CMP0177 policy

pamarcos committed:
Build the Google storage sample file on the CI

This is just to check that the CI builds ok.
TODO: revert this commit before merging

pamarcos committed:
Revert "Build the Google storage sample file on the CI"

This reverts commit 6c95be25ac1f6060ed43d266189b3d55cf65104e.

pamarcos commented:
> Maybe we can enable it for debug builds only? (it should build all examples)

That makes sense. I've created a separate PR to address that: https://github.com/ClickHouse/ClickHouse/pull/82905
I added the GCS sample just to confirm that my changes can link a binary that uses the SDK. Once I start working on the implementation, that example doesn't make any sense anymore, which is why I think it's not worth adding it into the repo since it'll get obsolete in the next PR.

azat commented on contrib/crc32c-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,123 @@
+  "${PROJECT_BINARY_DIR}/include/crc32c/crc32c_config.h"
+)
+
+include_directories("${PROJECT_BINARY_DIR}/include")
```
Does `crc32c_config.h` required by other public targets or only by the crc32c itself?

I guess `include_directories` will add it to `PUBLIC` list, which will add it also to all clickhouse targets, but maybe it is enough to have it in `PRIVATE`?

azat reviewed (commented):
@pamarcos haven't you forgot to include example (revert of revert) or you decided not to?

In general LGTM, but I think that crc32c cmake rules should be simplified.

pamarcos committed:
Revert "Revert "Build the Google storage sample file on the CI""

This reverts commit 80b3de5a2c5bf04a0cb558fed38316e6b190ec48.

pamarcos committed:
Clean up crc32c-cmake of unnecessary stuff

Originally it was a copy of a subset of crc32c's CMake.
I didn't want to change too much to ease syncing, but
let's tidy it up to keep unecessary stuff out.

pamarcos committed:
Merge branch 'master' into google-cloud-cpp-storage-build

pamarcos commented:
> haven't you forgot to include example (revert of revert) or you decided not to?

@azat I reverted the example on purpose because as it is right now we're not even building google-cloud-cpp anyway. It's only built for ClickHouse Cloud. I'll have to modify that in my next PR once we have something in public that actually uses the SDK, though.

I've brought back the example after simplifying the crc32c's CMake to ensure it's still working for all flavors, but my idea is to revert it once again once the CI is green.

>but I think that crc32c cmake rules should be simplified.

And you are right. I did it following your comments. Thanks for the thoughtful review on that.

azat commented:
> @azat I reverted the example on purpose because as it is right now we're not even building google-cloud-cpp anyway. It's only built for ClickHouse Cloud. I'll have to modify that in my next PR once we have something in public that actually uses the SDK, though.
>
>I've brought back the example after simplifying the crc32c's CMake to ensure it's still working for all flavors, but my idea is to revert it once again once the CI is green.

Yep, but what I meant is that it will be nice to have some binary that will use storage API to ensure that everything at least links correctly (since clickhouse binary does not use storage API AFAIU)

But I am not insist, I'm OK with your approach as well.

azat commented on CMakeLists.txt:
```diff
@@ -159,7 +159,7 @@ if (CMAKE_BUILD_TYPE_UC STREQUAL "DEBUG" OR CMAKE_BUILD_TYPE_UC STREQUAL "RELWIT
 
 option(ENABLE_TESTS "Provide unit_test_dbms target with Google.Test unit tests" ON)
-option(ENABLE_EXAMPLES "Build all example programs in 'examples' subdirectories" OFF)
+option(ENABLE_EXAMPLES "Build all example programs in 'examples' subdirectories" ON)
```
Debugging leftover?

pamarcos replied:
It was left on purpose to test the examples on the CI. Now that we're building all examples for AMD_DEBUG, this is not needed anymore. Also, I've added the examples of KMS and Storage to ensure they're built and linked correctly.

azat commented on contrib/crc32c-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,66 @@
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+)
+
+set_target_properties(_crc32c
+  PROPERTIES VERSION ${PROJECT_VERSION} SOVERSION ${PROJECT_VERSION_MAJOR})
```
I doubt we need `SOVERSION` for `*.a`, anyway `PROJECT_VERSION_MAJOR` will be likely used from ClickHouse top level cmake, which does not make sense

rschu1ze commented on .gitmodules:
```diff
@@ -372,3 +372,9 @@
 [submodule "contrib/chdig"]
 	path = contrib/chdig
 	url = https://github.com/azat/chdig.git
+[submodule "contrib/nlohmann-json"]
```
That will be the 3rd JSON parser in ClickHouse after RapidJSON and SIMDJson. Is it really needed?

rschu1ze replied:
Nah, not a tempting option.

pamarcos committed:
Merge branch 'master' into google-cloud-cpp-storage-build

pamarcos committed:
Add examples for google cloud storage and kms

pamarcos committed:
Split KMS and Storage builds

Use ENABLE_GOOGLE_CLOUD_CPP_KMS=1 to enable building KMS module
of Google Cloud Cpp.

pamarcos committed:
Simplify crc32c CMakeFile as per PR's review

pamarcos commented on contrib/crc32c-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,66 @@
+    "${CRC32C_DIR}/src/crc32c_read_le.h"
+    "${CRC32C_DIR}/src/crc32c_round_up.h"
+    "${CRC32C_DIR}/src/crc32c.cc"
+)
```
Done in 1872f700250cb5916bf8138dbca59aaafebe3c5e
Much cleaner now, thanks for the suggestion :ok_hand: 
At first I was trying to keep the CMakeFile.txt as close as possible to the original thinking it would ease maintenance, but it's true that for such a simple project we can create a much simpler file.

pamarcos committed:
Fix exclude for include quotes check

pamarcos committed:
Update flake.nix

pamarcos committed:
Fix build

pamarcos committed:
Build examples for all arch and OS

pamarcos committed:
Merge branch 'master' into google-cloud-cpp-storage-build

pamarcos committed:
Remove completely to make linter happy

pamarcos committed:
Fix protoc executable

pamarcos committed:
Fix something missed in the merge

pamarcos committed:
Fix tidy job for example

pamarcos committed:
WIP

pamarcos committed:
Add USE_GOOGLE_CLOUD_KMS

This allows us to build with KMS support separately.

pamarcos committed:
Clean google-cloud-cpp CMake files

pamarcos committed:
Build all targets when configuring with examples enabled

pamarcos commented on ci/jobs/build_clickhouse.py:
```diff
@@ -13,29 +13,29 @@
-    BuildTypes.RISCV64: f"      cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=None  -DENABLE_THINLTO=1 -DSANITIZE=          -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER={ToolSet.COMPILER_C} -DCMAKE_CXX_COMPILER={ToolSet.COMPILER_CPP} -DCOMPILER_CACHE=sccache -DENABLE_BUILD_PROFILING=1 -DENABLE_TESTS=0 -DENABLE_LEXER_TEST=0 -DE ...
-    BuildTypes.S390X: f"        cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=None  -DENABLE_THINLTO=1 -DSANITIZE=          -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER={ToolSet.COMPILER_C} -DCMAKE_CXX_COMPILER={ToolSet.COMPILER_CPP} -DCOMPILER_CACHE=sccache -DENABLE_BUILD_PROFILING=1 -DENABLE_TESTS=0 -DENABLE_LEXER_TEST=0 -DE ...
-    BuildTypes.LOONGARCH64: f"  cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=None  -DENABLE_THINLTO=0 -DSANITIZE=          -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER={ToolSet.COMPILER_C} -DCMAKE_CXX_COMPILER={ToolSet.COMPILER_CPP} -DCOMPILER_CACHE=sccache -DENABLE_BUILD_PROFILING=1 -DCMAKE_TOOLCHAIN_FILE={current_directory} ...
+    BuildTypes.RISCV64: f"      cmake --debug-trycompile -DCMAKE_VERBOSE_MAKEFILE=1 -LA -DCMAKE_BUILD_TYPE=None  -DENABLE_THINLTO=1 -DSANITIZE=          -DENABLE_CHECK_HEAVY_BUILDS=1 -DENABLE_CLICKHOUSE_SELF_EXTRACTING=1 -DCMAKE_C_COMPILER={ToolSet.COMPILER_C} -DCMAKE_CXX_COMPILER={ToolSet.COMPILER_CPP} -DCOMPILER_CACHE=sccache -DENABLE_BUILD_PROFILING=1 -DENABLE_TESTS=0 -DENABLE_LEXER_TEST=0 -DE ...
```
Since I added one example for KMS and another one for Storage, I want to make sure they can be built and linked against in every flavor we have. I'll remove that once the PR is ready to be reviewed again. Which I reminds me that even though I warned Azat, I didn't move it to draft.

pamarcos replied:
I added some TODO comments to remind myself of removing that before merging in https://github.com/ClickHouse/ClickHouse/pull/82881/commits/d24fd0365e2580b04123d18451a61ab792816320

pamarcos committed:
Address comments from PR review

pamarcos commented on contrib/crc32c-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+target_include_directories(_crc32c
+  PUBLIC
+    $<BUILD_INTERFACE:${CRC32C_DIR}/include>
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
```
You're right: https://github.com/ClickHouse/ClickHouse/pull/82881/commits/d24fd0365e2580b04123d18451a61ab792816320

pamarcos commented on contrib/CMakeLists.txt:
```diff
@@ -242,6 +242,8 @@ add_contrib (prometheus-protobufs-cmake prometheus-protobufs prometheus-protobuf
 add_contrib (numactl-cmake numactl)
 
+add_contrib (nlohmann-json-cmake nlohmann-json)
+add_contrib (crc32c-cmake crc32c)
```
Added in https://github.com/ClickHouse/ClickHouse/pull/82881/commits/d24fd0365e2580b04123d18451a61ab792816320

pamarcos committed:
Format google-cloud-cpp sample to comply with our style

pamarcos commented on ci/jobs/scripts/check_style/check_cpp.sh:
```diff
@@ -14,7 +14,7 @@
 LC_ALL="en_US.UTF-8"
 ROOT_PATH=$(git rev-parse --show-toplevel)
-EXCLUDE='build/|integration/|widechar_width/|glibc-compatibility/|poco/|memcpy/|consistent-hashing|benchmark|tests/.*.cpp|programs/keeper-bench/example.yaml|base/base/openpty.h'
+EXCLUDE='build/|integration/|widechar_width/|glibc-compatibility/|poco/|memcpy/|consistent-hashing|benchmark|tests/.*.cpp|programs/keeper-bench/example.yaml|base/base/openpty.h|examples/gcloud.*'
```
Done at https://github.com/ClickHouse/ClickHouse/pull/82881/commits/0a5606f3f400ef43aad998d809a405a0237c809e

pamarcos committed:
Add ci-build label to just build binaries

pamarcos committed:
Revert "WIP"

This reverts commit fec94068e920d794a8f6bee95210ed4et pe90c4530.

pamarcos committed:
Don't error in case gRPC or protobuf are not enabled

pamarcos committed:
Disable building examples for non-supported Google Cloud Cpp platforms

pamarcos committed:
Merge branch 'master' into google-cloud-cpp-storage-build

rschu1ze commented on contrib/crc32c-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+set(CRC32C_DIR ${ClickHouse_SOURCE_DIR}/contrib/crc32c)
+
+configure_file(
```
Right, we could have such a file per architecture and select the right one in CMake based on the target architecture. As an example, see "contrib/openssl-cmake" which follows similar logic.

pamarcos replied:
I did it in https://github.com/ClickHouse/ClickHouse/pull/82881/commits/e922e457ddbcd258ce6f6c039069382cac07a0fe
I think we just need aarch64 vs x86_64 and within the latter sse42 vs no_sse42

pamarcos committed:
Commit static versions of crc32c_config

pamarcos committed:
Merge branch 'master' into google-cloud-cpp-storage-build

pamarcos committed:
Rename nlohmann's JSON in AI SDK to avoid clash

pamarcos committed:
Restore original build_clickhouse.py

pamarcos commented:
@azat, @rschu1ze please take another look whenever you can.
In https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=82881&sha=d40f814abea63affd58b3e048b196c9976e3c78f&name_0=PR there's the proof that `gcloud_storage` builds and links successfully for Linux, macOS (x86_64 and aarch64) and FreeBSD.

rschu1ze commented on contrib/crc32c-cmake/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+set(CRC32C_DIR ${ClickHouse_SOURCE_DIR}/contrib/crc32c)
+set(CRC32C_CMAKE_DIR ${ClickHouse_SOURCE_DIR}/contrib/crc32c-cmake)
+
+configure_file(
```
We don't need l. 4-7 anymore, no?

rschu1ze commented on contrib/google-cloud-cpp-cmake/CMakeLists.txt:
```diff
@@ -11,13 +11,28 @@ if(NOT ENABLE_GOOGLE_CLOUD_CPP)
+
 if(NOT ENABLE_GRPC)
-    message (${RECONFIGURE_MESSAGE_LEVEL} "Can't use Google Cloud Cpp without gRPC")
+    message(STATUS "Can't use Google Cloud Cpp without gRPC")
```
Guess this should be a FATAL_ERROR. Silently not building gcp despite the user requested it is worse.

pamarcos replied:
I modified it so that it doesn't throw a failure when configuring the flavors: riscv64, s390x, loongarch64, ppc64le.
What I'm going to do is to explicitly enable `GOOGLE_CLOUD_CPP` only for those flavors we know it works in: Linux, macOS and FreeBSD, ok?

rschu1ze replied:
Yepp, tnx

rschu1ze commented on contrib/google-cloud-cpp-cmake/CMakeLists.txt:
```diff
@@ -11,13 +11,28 @@ if(NOT ENABLE_GOOGLE_CLOUD_CPP)
-if (NOT ENABLE_PROTOBUF)
-    message( ${RECONFIGURE_MESSAGE_LEVEL} "Can't use Google Cloud Cpp without protobuf")
+if(NOT ENABLE_PROTOBUF)
+    message(STATUS "Can't use Google Cloud Cpp without protobuf")
```
same^^

pamarcos commented on ci/jobs/scripts/check_style/check_cpp.sh:
```diff
@@ -353,6 +353,7 @@ QUOTE_EXCLUSIONS=(
     --exclude "$ROOT_PATH/utils/memcpy-bench/glibc/*"
 )
 find $ROOT_PATH/{src,programs,utils} -name '*.h' -or -name '*.cpp' | \
+  grep -vP $EXCLUDE |
```
Yeah, agree with the `check_cpp.sh` changes, but I needed the other two changes to avoid waiting ad infinitum to build the macOS and FreeBSD flavors

pamarcos committed:
Address comments from PR review

pamarcos committed:
Revert "Restore original build_clickhouse.py"

This reverts commit fe6e511a733a65ccf0448ef14821e4249b5922d3.

pamarcos commented on contrib/CMakeLists.txt:
```diff
@@ -242,7 +242,9 @@ add_contrib (prometheus-protobufs-cmake prometheus-protobufs prometheus-protobuf
 add_contrib (numactl-cmake numactl)
 
-add_contrib (google-cloud-cpp-cmake google-cloud-cpp) # requires grpc, protobuf, absl
+add_contrib (nlohmann-json-cmake nlohmann-json)
```
Done in https://github.com/ClickHouse/ClickHouse/pull/82881/commits/522fc03bb8ff6184fc31373616e680db22714b6e

pamarcos committed:
fixup! Address comments from PR review

pamarcos committed:
Merge branch 'master' into google-cloud-cpp-storage-build

pamarcos committed:
Fix crc32c header include dir

pamarcos committed:
Only build Google Cloud Cpp for x86_64 and aarch64

pamarcos committed:
Revert "Revert "Restore original build_clickhouse.py""

This reverts commit a3cae01ef319b8eb32145a202682e5250fc262d4.

pamarcos committed:
Disable Google Cloud Cpp for arm_v80compat

| /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/contrib/crc32c/src/crc32c_arm64.cc:87:20: error: always_inline function 'vmull_p64' requires target feature 'aes', but would be inlined into function 'ExtendArm64' that is compiled without support for 'aes'
|    87 |     t1 = (uint64_t)vmull_p64(crc1, k1);
|       |                    ^
| /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/contrib/crc32c/src/crc32c_arm64.cc:88:20: error: always_inline function 'vmull_p64' requires target feature 'aes', but would be inlined into function 'ExtendArm64' that is compiled without support for 'aes'
|    88 |     t0 = (uint64_t)vmull_p64(crc0, k0);
|       |                    ^

pamarcos commented on contrib/google-cloud-cpp-cmake/CMakeLists.txt:
```diff
@@ -76,29 +99,37 @@ set(PROJECT_SOURCE_DIR ${PROJECT_SOURCE_DIR_BAK})
+    )
+
+    # The library is large - avoid bloat.
+    if(OMIT_HEAVY_DEBUG_SYMBOLS)
+        target_compile_options(_gcloud PRIVATE -g0)
+    endif()
```
I missed it! Fixed in https://github.com/ClickHouse/ClickHouse/pull/84100
